monai/losses/segcalib.py

```diff
@@ -0,0 +1,173 @@
+            raise ValueError("Kernel ops must be either mean or gaussian")
+
+        if distance_type not in ["l1", "l2"]:
+            raise ValueError("Distance type must be either L1 or L2")
+
+        self.kernel_ops = kernel_ops
+        self.distance_type = distance_type
+        self.alpha = alpha
+
+        self.nc = classes
+        self.ks = kernel_size
+        self.cross_entropy = nn.CrossEntropyLoss()
+
+        if kernel_ops == "gaussian":
+            self.svls_layer = GaussianFilter(ksize=kernel_size, sigma=sigma, channels=classes)
+
+        self.old_pt_ver = not pytorch_after(1, 10)
+
+    def ce(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """
+        Compute CrossEntropy loss for the input logits and target.
+        Will remove the channel dim according to PyTorch CrossEntropyLoss:
+        https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html?#torch.nn.CrossEntropyLoss.
+
+        """
+        n_pred_ch, n_target_ch = input.shape[1], target.shape[1]
+        if n_pred_ch != n_target_ch and n_target_ch == 1:
+            target = torch.squeeze(target, dim=1)
+            target = target.long()
+        elif self.old_pt_ver:
+            warnings.warn(
+                f"Multichannel targets are not supported in this older Pytorch version {torch.__version__}. "
+                "Using argmax (as a workaround) to convert target to a single channel."
+            )
+            target = torch.argmax(target, dim=1)
+        elif not torch.is_floating_point(target):
+            target = target.to(dtype=input.dtype)
+
+        return self.cross_entropy(input, target)  # type: ignore[no-any-return]
+
+    def get_constr_target(self, mask: torch.Tensor) -> torch.Tensor:
+        mask = mask.unsqueeze(1)  # unfold works for 4d.
+
+        bs, _, h, w = mask.shape
+        unfold = torch.nn.Unfold(kernel_size=(self.ks, self.ks), padding=self.ks // 2)
+
+        rmask = []
+
+        if self.kernel_ops == "mean":
+            umask = unfold(mask.float())
+
+            for ii in range(self.nc):
+                rmask.append(torch.sum(umask == ii, 1) / self.ks**2)
+
+        if self.kernel_ops == "gaussian":
+            oh_labels = (
+                F.one_hot(mask[:, 0].to(torch.int64), num_classes=self.nc).contiguous().permute(0, 3, 1, 2).float()
+            )
+            rmask = self.svls_layer(oh_labels)
+
+            return rmask
+
+        rmask = torch.stack(rmask, dim=1)
+        rmask = rmask.reshape(bs, self.nc, h, w)
+
+        return rmask
+
+    def forward(self, inputs: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
+        loss_ce = self.ce(inputs, targets)
+
+        utargets = self.get_constr_target(targets)
+
+        if self.distance_type == "l1":
+            loss_conf = torch.abs(utargets - inputs).mean()
+
+        if self.distance_type == "l2":
+            loss_conf = (torch.abs(utargets - inputs) ** 2).mean()
```

Comment:
```suggestion
        if self.distance_type == "l1":
            loss_conf = utargets.sub(inputs).abs_().mean()
        elif self.distance_type == "l2":
            loss_conf = utargets.sub(inputs).pow_(2).abs_().mean()
```
Maybe a little faster?